third_party/nvfuser/csrc/scheduler/registry.cpp

```diff
@@ -2038,10 +2044,272 @@ class PersistentKernelScheduler : public SchedulerEntry {
   }
 };
 
+class MatmulScheduler : public SchedulerEntry {
+ public:
+  explicit MatmulScheduler(
+      Fusion* fusion,
+      SchedulerRuntimeInfo& runtime_info,
+      HeuristicSummary* data_cache = nullptr)
+      : SchedulerEntry(ScheduleHeuristic::Matmul) {
+    computeHeuristics(fusion, runtime_info);
+  }
+
+  void schedule(Fusion* fusion) override {
+    FUSER_PERF_SCOPE("Schedule Matmul Fusion");
+    scheduleMatmul(fusion, matmulParams());
+  }
+
+  static bool canScheduleCompileTime(Fusion* fusion) {
+    // Check that inputs of all select/gather-like ops are fusion inputs
+    if (rejectScheduleForSelectLikeOps(fusion, ScheduleHeuristic::Matmul)) {
+      return false;
+    }
+
+    // The plan:
+    // 1. check if MmaOp is defined
+    // 2. check if TV bound to MmaOp is in outputs
+    // 3. check if inputs for output are in local memory
+    // 4. check if local_mem of MmaOp have supported layout
+    // 5. [TODO] check if local_mem of MmaOp are a result of broadcast
+    //    on fusion inputs
+
+    // #1
+    std::vector<Expr*> mmaExprs;
+    {
+      // TODO: outline to ir_utils namespace
+      for (const auto expr : fusion->exprs()) {
+        if (expr->isA<MmaOp>()) {
+          mmaExprs.push_back(expr);
+        }
+      }
+    }
+
+    if (mmaExprs.empty()) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          ScheduleHeuristic::Matmul, "No mma op to schedule");
+      return false;
+    }
+
+    // #2
+    // NOTE: this may change in future, with support for C matrix and beta
+    {
+      // TODO: outline to ir_utils namespace
+      const auto& outputs = fusion->outputs();
+      std::vector<Expr*> mmaOutputs;
+
+      for (auto expr : mmaExprs) {
+        const auto& exprOutputs = expr->outputs();
+        if (exprOutputs.size() != 1) {
+          // we are interested only in mma ops with single output val
+          continue;
+        }
+        if (std::find(outputs.cbegin(), outputs.cend(), exprOutputs.front()) !=
+            outputs.cend()) {
+          mmaOutputs.push_back(expr);
+        }
+      }
```

Comment:
You can just check ->isFusionOutput/Input directly on the tensor views.

Reply:
With how we do matmul at the moment ([link](https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/test/test_utils.cpp#L253)) this check will not help us. The MmaOp direct inputs are a result of already processed fusion inputs which are marked to be in local memory, to answer @naoyam question why this is not in global memory.

I'm thinking about tracking down if MmaOP inputs are a result of some operations on fusion input - this might be helpful in future if we would like to add support for conversion (e.g. fp32 -> tf32).
It is similar situation for MmaOp output, with prepared tests it is indeed an output of fusion but if we were to support full alpha:

```
C = alpha*AxB
```
then the result of MmaOP will be a partial data, used to multiply by alpha to produce the final result.

To have it done I will need to build a 'map' with relationships (producer-consumer) between Vals with MmaOp in the center of this map (I boldly assume there is a single MmaOp in the fusion).

My major concern is that this map will have a fixed 'routes' for predefined set of expressions. On the other hand, defining checks for specific patterns/routes in the Vals' dependencies may be useful for adding extended functionality (mentioned alpha or beta with additional input tensor, later also epilogues).

An example of such a route could be
- for vanilla matmul (things to be checked):
-- MmaOP output is fusion output,
-- MmaOp inputs are a result of adding broadcast to the input TVs
-- fusion has two inputs,
-- fusion has one output,
- for alpha matmul:
-- fusion has two inputs,
-- fusion has one output,
-- fusion output is a consumer of product of MmaOp output (matmul result) and 3rd fusion input (alpha)
-- MmaOp inputs are a result of adding broadcast to input TVs,

Does it make sense? There is a risk that such a pattern matching might be expensive, even for compile time processing.

Reply:
I don't think this should be expensive. I guess maybe the checks though I was thinking (maybe not immediately) would be:
-- Broadcast pattern of MmaOp (I think we can just check where the broadcast positions are relative to the two inputs of the MMA Op)
-- All operations on inputs to MmaOp are Unary ops
-- All operations on outputs of MmaOp are Unary ops

I think unary ops should be quite straightforward to process in the Mma scheduler, but maybe that's not the case?